Adds ability to mark parameters as multi-value

weierophinney opened the pull request:
In symfony/console, you can provide a "mode" to options, which is generally one of "none" or "required". However, you can also mask "required" values with a mode indicating the option is an array; in this scenario, the user can supply the option multiple times, and the aggregate is presented as an array when the command retrieves the option value.

Since parameters are essentially options with the ability to prompt for values if they were not provided, this patch adds support for that mode.

The support is added at the `AbstractParamAwareInput` level, when calling `getParam()`. The class now has logic to determine if the mode indicates an array of values, and if so, it will then prompt repeatedly until an empty line (i.e., user presses "return") is provided. If the value is not required, the user can press "return" immediately; otherwise, they will be prompted until at least one value is provided.

The various questions created have been updated to provide information to the user indicating that multiple values may be entered, and also to let them know when a value MUST be entered (i.e., parameter is required).

I have tests for all functionality, which were guided by actual usage within an application (where I was able to determine what did and did not work.)

Fixes #23

weierophinney commented on src/Input/AbstractParamAwareInput.php:
```diff
@@ -167,4 +169,110 @@ public function getStream()
+        $validator = $question->getValidator();
+        $value     = null;
+        $values    = [];
+        do {
+            if (null !== $value) {
+                $values[] = $value;
+            }
+            $value = $this->helper->ask($this, $this->output, $question);
+
+            if ($valueIsRequired && [] === $values) {
+                $question->setValidator(function ($value) use ($validator) {
+                    if (null === $value || '' === $value) {
+                        return $value;
+                    }
+
+                    return $validator($value);
+                });
+            }
+        } while (! in_array($value, [null, ''], true));
```
This is the "magic" that allows multi-prompting. Essentially, we prompt until we do not get a value. However, to ensure that we don't attempt to validate an empty line, I need to check for that after the first submission, so I prepend a validator for empty lines to the existing validator (but only  after the first iteration through the loop). Afterwards, I reset the validator.

In this way, we'll keep prompting until the user stops submitting values.

weierophinney commented on src/Input/ChoiceParam.php:
```diff
@@ -33,12 +36,21 @@ public function getQuestion(): Question
+        $multiPrompt   = sprintf(
+            "\n(Multiple selections allowed; hit Return after each.%s Hit Return to stop prompting)\n",
+            $this->isRequired() ? ' At least one selection is required.' : ''
+        );
```
One thing I noted when playing with the `ChoiceQuestion`: it does not present any information to the user whatsoever that they can (a) submit multiple values, and (b) _how_ they should submit them (the `ChoiceQuestion` normally wants them comma-separated when it is a `multiSelect`).

I've added verbiage here and in the `StandardQuestionTrait` around the ability to submit multiple values, and whether at least one is required.

samsonasik commented on src/Input/AbstractParamAwareInput.php:
```diff
@@ -82,29 +87,26 @@ public function __construct(InputInterface $input, OutputInterface $output, Ques
-            $normalizer = $question->getNormalizer();
-
-            return $normalizer === null ? $value : $normalizer($value);
+        $valueIsArray = (bool) ($inputParam->getOptionMode() & InputOption::VALUE_IS_ARRAY);
```
`$valueIsArray` will always be false ? checked like this:

```php
➜  ~ php -a
Interactive shell

php > var_dump((bool) (false & 8));
bool(false)

php > var_dump((bool) (true & 8));
bool(false)

php > var_dump((bool) (1 & 8));
bool(false)

php > var_dump((bool) (0 & 8));
bool(false)
```

ref https://3v4l.org/bhFWC

michalbundyra replied:
@samsonasik but you forgot to check with the following values:

```php
var_dump((bool) (8 & 8)); // true
var_dump((bool) (9 & 8)); // true
```

These are interesting cases for us, so when parameter is set to have multiple values.

weierophinney replied:
@samsonasik Bitwise operators return an integer. In this particular case, `getOptionMode()` will return an integer, and we are using `&` to determine which bits it shares in common with `InputOption::VALUE_IS_ARRAY`, if any. This will return `0` if there are no bits in common, and an integer greater than or equal to `1` if there are. As such, casting to boolean is used to determine if the current option mode includes that bit.

By **default**, none of the various elements includes it. However, you can use `setOptionMode()` to set a different mask of bits, such as `InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY`. In that case, this operation will evaluate to `true`.

A few of the tests I've provided demonstrate this, as they will only work correctly if `VALUE_IS_ARRAY` is in the option mode bitmask.

samsonasik replied:
ok 👍

weierophinney commented:
> In general it looks fine. I do not like only one thing:
> 
>     * method `setOptionMode` should not be available on our input params, imo.
> 
> 
> How I see it is to have additional setter: `setMultipleValues(bool true)` or something like that which will add the flag internally.
> 
> It should be added per input param type - so for example, we should not have this setter in `BoolParam` as it is using `InputOption::VALUE_NONE` (yes/no question).
> 
> As far as I can see it is the only place where we are mixing (in visible way for users) input params with input options.

When writing the tests, I was having a little nagging feeling that allowing users to set the option mode however they want was a potential issue, despite the safeguards you had me put in. I like this idea a lot better, and will start working on it today.

weierophinney commented:
@michalbundyra Just pushed the changes as requested. I've removed `setOptionMode()` from both the `InputParamInterface` and `AbstractInputParam`, and added the trait `AllowMultipleTrait`, which will change the mask of `$optionMode` based on the flag provided. This trait is now used by all param implementations EXCEPT the `BoolParam`.

I've updated docs as well.
